README
Dancer-Plugin-Auth-Twitter

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Dancer::Plugin::Auth::Twitter

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Dancer-Plugin-Auth-Twitter

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Dancer-Plugin-Auth-Twitter

    CPAN Ratings
        http://cpanratings.perl.org/d/Dancer-Plugin-Auth-Twitter

    Search CPAN
        http://search.cpan.org/dist/Dancer-Plugin-Auth-Twitter/


LICENSE AND COPYRIGHT

Copyright (C) 2010 Alexis Sukrieh

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Dancer-Plugin-Auth-Twitter

{{$NEXT}}
 [API CHANGES]

 [BUG FIXES]

 [DOCUMENTATION]

 [ENHANCEMENTS]

 [NEW FEATURES]

 [STATISTICS]

0.07 2015-02-03
 [DOCUMENTATION]
 - Add note about 'can't verify ssl peer'. (GH#16)

 [ENHANCEMENTS]
 - Replace default engine to be Net::Twitter::Lite. (GH#15, odyniec)

 [MISC]
 - Add tests. (GH#17, odyniec)

 [STATISTICS]
 - code churn: 9 files changed, 190 insertions(+), 53 deletions(-)

0.06 2014-02-15
 [BUG FIXES]
 - Net::Twitter now needs to have 'SSL => 1'.  (GH#10, reported by berekuk)

0.05 2013-06-06
 [MISC]
 - Add 'Net::Twitter >= 4.0001' to the dependency list (for v1.1 of the
   REST API). (Yanick Champoux)

0.04 2013-06-06
 [BUG FIXES]
 - Was using 'core()', which is not exported. GH#8 (Vyacheslav Matyukhin)

 [ENHANCEMENTS]
 - Use v1.1 of the Twitter API REST (v1 is phased out by Twitter  on June
   11th). GH#7 (Vyacheslav Matyukhin)

0.03 2013-05-14
 [BUG FIXES]
 - fix L<> markup in POD. GH#5 (Andreas Koenig, Yanick Champoux)

 [ENHANCEMENTS]
 - Auth denied by user now redirects to $callback_fail. GH#2 (Will Orr) *
   support for both authentication and authorization method. GH#3 (Simon
   Elsbrock)
 - Change 'warn' for 'debug'. GH#4 (berekuk)

0.02 2011-01-17
 [ENHANCEMENTS]
 - allow customization of callback redirection URIs. (Yanick Champoux)

 [MISC]
 - Some release tweaks. (Alexis Sukrieh)

0.01 2010-12-20
 - First version, release as part of the Perl Dancer Advent Calendar 2010



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.012.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Authenticate with Twitter",
  "dist_author" => [
    "Alexis Sukrieh <sukria\@sukria.net>",
    "Dancer Core Developers"
  ],
  "dist_name" => "Dancer-Plugin-Auth-Twitter",
  "dist_version" => "0.06",
  "license" => "perl",
  "module_name" => "Dancer::Plugin::Auth::Twitter",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Dancer" => 0,
    "Dancer::Plugin" => 0,
    "Net::Twitter" => "4.00001",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "List::Util" => 0,
    "Test::More" => 0,
    "perl" => "5.006"
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::Spec::Functions" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "List::Util" => 0,
  "Module::Build" => "0.3601",
  "Test::More" => 0,
  "perl" => "5.006"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Dancer/Plugin/Auth/Twitter.pm
package Dancer::Plugin::Auth::Twitter;
#ABSTRACT: Authenticate with Twitter

use strict;
use warnings;

use Dancer ':syntax';
use Dancer::Plugin;
use Carp 'croak';
use Class::Load qw/ load_class /;
use Net::Twitter::Lite::WithAPIv1_1 0.12006;

# Net::Twitter singleton, accessible via 'twitter'
our $_twitter;
sub twitter { $_twitter }
register 'twitter' => \&twitter;

# init method, to create the Net::Twitter object
my $consumer_key;
my $consumer_secret;
my $callback_url;
my $callback_success;
my $callback_fail;

my $engine;

register 'auth_twitter_init' => sub {
    my $config = plugin_setting;

    $consumer_secret = $config->{consumer_secret};
    $consumer_key    = $config->{consumer_key};
    $callback_url    = $config->{callback_url};

    $callback_success = $config->{callback_success} || '/';
    $callback_fail    = $config->{callback_fail}    || '/fail';

    for my $param (qw/consumer_key consumer_secret callback_url/) {
        croak "'$param' is expected but not found in configuration" 
            unless $config->{$param};
    }

    debug "new twitter with $consumer_key , $consumer_secret, $callback_url";

    $engine = $config->{engine} || 'Net::Twitter::Lite::WithAPIv1_1';

    $engine .= '::WithAPIv1_1' if $engine eq 'Net::Twitter::Lite';

    die "engine must be 'Net::Twitter' or 'Net::Twitter::Lite::WithAPIv1_1': '$engine' not supported\n"
        unless grep { $engine eq $_ } qw/ Net::Twitter Net::Twitter::Lite::WithAPIv1_1 /;

    $_twitter = load_class($engine)->new(
        ( traits => [ qw/ API::RESTv1_1 OAuth / ] ) x ( $engine eq 'Net::Twitter' ),
        'consumer_key'      => $consumer_key, 
        'consumer_secret'   => $consumer_secret,
         ssl                 => 1,
    );

};

# define a route handler that bounces to the OAuth authorization process
register 'auth_twitter_authorize_url' => sub {
    if (not defined twitter) {
        croak "auth_twitter_init must be called first";
    }

    my $uri = twitter->get_authorization_url( 
        'callback' => $callback_url
    );

    session request_token        => twitter->request_token;
    session request_token_secret => twitter->request_token_secret;
    session access_token         => '';
    session access_token_secret  => '';

    debug "auth URL : $uri";
    return $uri;
};

# define a route handler that bounces to the OAuth authentication process
register 'auth_twitter_authenticate_url' => sub {
    if (not defined twitter) {
        croak "auth_twitter_init must be called first";
    }

    my $uri = twitter->get_authentication_url(
        'callback' => $callback_url
    );

    session request_token        => twitter->request_token;
    session request_token_secret => twitter->request_token_secret;
    session access_token         => '';
    session access_token_secret  => '';

    debug "auth URL : $uri";
    return $uri;
};

get '/auth/twitter/callback' => sub {

    debug "in callback...";

    # A user denying access should be considered a failure
    return redirect $callback_fail if (params->{'denied'});

    if (   !session('request_token')
        || !session('request_token_secret')
        || !params->{'oauth_verifier'})
    {
        return send_error 'no request token present, or no verifier';
    }

    my $token               = session('request_token');
    my $token_secret        = session('request_token_secret');
    my $access_token        = session('access_token');
    my $access_token_secret = session('access_token_secret');
    my $verifier            = params->{'oauth_verifier'};

    if (!$access_token && !$access_token_secret) {
        twitter->request_token($token);
        twitter->request_token_secret($token_secret);
        ($access_token, $access_token_secret) = twitter->request_access_token('verifier' => $verifier);

        # this is in case we need to register the user after the oauth process
        session access_token        => $access_token;
        session access_token_secret => $access_token_secret;
    }

    # get the user
    twitter->access_token($access_token);
    twitter->access_token_secret($access_token_secret);

    my $twitter_user_hash;
    eval {
        $twitter_user_hash = twitter->verify_credentials();
    };

    if ($@ || !$twitter_user_hash) {
        Dancer::Logger::core("no twitter_user_hash or error: ".$@);
        return redirect $callback_fail;
    }

    $twitter_user_hash->{'access_token'} = $access_token;
    $twitter_user_hash->{'access_token_secret'} = $access_token_secret;

    # save the user
    session 'twitter_user'                => $twitter_user_hash;
    session 'twitter_access_token'        => $access_token,
    session 'twitter_access_token_secret' => $access_token_secret,

    redirect $callback_success;
};
 
register_plugin;

__END__


=head1 SYNOPSIS

    package SomeDancerApp;
    use Dancer ':syntax';
    use Dancer::Plugin::Auth::Twitter;

    auth_twitter_init();

    before sub {
        if (not session('twitter_user')) {
            redirect auth_twitter_authenticate_url;
        }
    };

    get '/' => sub {
        "welcome, ".session('twitter_user')->{'screen_name'};
    };

    get '/fail' => sub { "FAIL" };

    ...

=head1 CONCEPT

This plugin provides a simple way to authenticate your users through Twitter's
OAuth API. It provides you with a helper to build easily a redirect to the
authentication URI, defines automatically a callback route handler and saves the
authenticated user to your session when done.

=head1 PREREQUESITES

In order for this plugin to work, you need the following:

=over 4 

=item * Twitter application

Anyone can register a Twitter application at L<http://dev.twitter.com/>. When
done, make sure to configure the application as a I<Web> application.

=item * Configuration

You need to configure the plugin first: copy your C<consumer_key> and C<consumer_secret> 
(provided by Twitter) to your Dancer's configuration under
C<plugins/Auth::Twitter>:

    # config.yml
    ...
    plugins:
      "Auth::Twitter":
        consumer_key:     "1234"
        consumer_secret:  "abcd"
        callback_url:     "http://localhost:3000/auth/twitter/callback"
        callback_success: "/"
        callback_fail:    "/fail"
        engine:           Net::Twitter::Lite::WithAPIv1_1

C<callback_success> and C<callback_fail> are optional and default to 
'/' and '/fail', respectively.

The engine is optional as well. The supported engines are
C<Net::Twitter> and C<Net::Twitter::Lite::WithAPIv1_1> (C<Net::Twitter::Lite> can also 
be used as a synonym for the latter).
By default the plugin will use L<Net::Twitter::Lite::WithAPIv1_1>.

Note that you also need to provide your callback url, whose route handler is automatically
created by the plugin.

=item * Session backend

For the authentication process to work, you need a session backend, in order for
the plugin to store the authenticated user's information.

Use the session backend of your choice, it doesn't make a difference, see
L<Dancer::Session> for details about supported session engines, or
L<search the CPAN for new ones|http://search.cpan.org/search?query=Dancer-Session>.

=back

=head1 EXPORT

The plugin exports the following symbols to your application's namespace:

=head2 twitter

The plugin uses a L<Net::Twitter> or L<Net::Twitter::Lite::WithAPIv1_1> 
object to do its job. You can access this
object with the C<twitter> symbol, exported by the plugin.

=head2 auth_twitter_init

This function should be called before your route handlers, in order to
initialize the underlying L<Net::Twitter> or L<Net::Twitter::Lite::WithAPIv1_1> 
object. 

=head2 auth_twitter_authorize_url

This function returns an authorize URI for redirecting unauthenticated users.
You should use this in a before filter like the following:

    before sub {
        # we don't want to bounce for ever!
        return if request->path =~ m{/auth/twitter/callback};
    
        if (not session('twitter_user')) {
            redirect auth_twitter_authorize_url();
        }
    };

When the user authenticate with Twitter's OAuth interface, she's going to be
bounced back to C</auth/twitter/callback>.

=head2 auth_twitter_authenticate_url

Similar to auth_twitter_authorize_url, but this function instead returns an
authenticate instead of authorize URI for redirecting unauthenticated users,
which results in a slightly different behaviour.

See L<https://dev.twitter.com/pages/sign_in_with_twitter|here> to learn about
the differences.

=head1 ROUTE HANDLERS

The plugin defines the following route handler automatically

=head2 /auth/twitter/callback

This route handler is responsible for catching back a user that has just
authenticated herself with Twitter's OAuth. The route handler saves tokens and
user information in the session and then redirects the user to the URI
specified by C<callback_success>.  

If the validation of the token returned by Twitter failed or was denied,
the user will be redirect to the URI specified by C<callback_fail>.

=head1 TIPS AND TRICKS

If you get the error

    Net::Twitter::Role::OAuth::get_authorization_url(): 
        GET https://api.twitter.com/oauth/request_token failed: 500 
        Can't verify SSL peers without knowing which Certificate Authorities to trust

you can silence it by setting the environment variable C<PERL_LWP_SSL_VERIFY_HOSTNAME> to C<0>.

=head1 ACKNOWLEDGEMENTS

This plugin has been written as a port of
L<Catalyst::Authentication::Credential::Twitter> written by 
Jesse Stay.
  
This plugin was part of the Perl Dancer Advent Calendar 2010.


=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.012.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Authenticate with Twitter",
  "AUTHOR" => "Alexis Sukrieh <sukria\@sukria.net>, Dancer Core Developers",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Dancer-Plugin-Auth-Twitter",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Dancer::Plugin::Auth::Twitter",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Dancer" => 0,
    "Dancer::Plugin" => 0,
    "Net::Twitter" => "4.00001",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "List::Util" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.06",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Dancer" => 0,
  "Dancer::Plugin" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::Spec::Functions" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "List::Util" => 0,
  "Module::Build" => "0.3601",
  "Net::Twitter" => "4.00001",
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



